src/recordlinker/database/mpi_service.py

```diff
@@ -263,16 +264,51 @@ def get_patients_by_reference_ids(
     return [patients_by_id.get(ref_id) for ref_id in reference_ids]
 
 
+def get_patients_by_person_ids(
+    session: orm.Session, person_ids: typing.Sequence[int]
+) -> list[models.Patient | None]:
+    """
+    Retrieve all the Patients by their person ids. If no Patient is found for a
+    person_id, no data will be returned.
+    """
+    # Check that the person_ids are valid
+    if not all(isinstance(pid, int) for pid in person_ids):
+        raise SQLAlchemyError("All person_ids must be integers.")
+
+    # Get the Patients by person_id
+    query = select(models.Patient).where(models.Patient.person_id.in_(person_ids))
+    patients_by_person_id = {
+        patient.person_id: patient for patient in session.execute(query).scalars().all()
+    }
+    return [
+        patient
+        for patient in (patients_by_person_id.get(person_id) for person_id in person_ids)
+        if patient is not None
+    ]
+
+
 def get_person_by_reference_id(
-    session: orm.Session, reference_id: uuid.UUID
+    session: orm.Session, person_reference_id: uuid.UUID
 ) -> models.Person | None:
     """
-    Retrieve the Person by their reference id
+    Retrieve a single Person by their person reference ID.
     """
-    query = select(models.Person).where(models.Person.reference_id == reference_id)
+    query = select(models.Person).where(models.Person.reference_id == person_reference_id)
     return session.scalar(query)
 
 
+def get_persons_by_reference_ids(
+    session: orm.Session, person_reference_ids: typing.Sequence[uuid.UUID]
+) -> typing.Sequence[models.Person] | None:
+    """
+    Retrieve multiple Persons by their person reference IDs.
+    """
+    if not person_reference_ids:
+        return None
+    query = select(models.Person).where(models.Person.reference_id.in_(person_reference_ids))
+    return session.scalars(query).all()
```

Comment:
Let's say we call this method with two person_reference_ids, a valid one (meaning this UUID matches some person cluster in the DB) and an invalid one (its a UUID, but doesn't match any person cluster in the DB).

In that situation, you will get back a list of person objects, its just the list will be of length 1 not 2.  I believe that's going to cause problems for the router call, because its only raising a 422 if the value is None.

Reply:
I hadn't thought of that scenario. I added back in a 422 check for getting the person_reference_ids which should raise a 422 if any of the passed person_reference_ids are invalid. It seems safer to fail in this scenario, though I'm curious if NBS has provided any insight about how they would prefer to handle that scenario; should we proceed with a merge when possible or fail if any of the person reference ids are invalid?